nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1239,6 +1318,86 @@ impl<'a> Planner<'a> {
+        let source_repo = if self.is_zone_ready_for_update(zone_kind)? {
+            self.input.tuf_repo().description()
+        } else {
+            self.input.old_repo().description()
+        };
+        source_repo.zone_image_source(zone_kind)
+    }
+
+    /// Return `true` iff a zone of the given kind is ready to be updated;
+    /// i.e., its dependencies have been updated.
+    fn is_zone_ready_for_update(
+        &self,
+        zone_kind: ZoneKind,
+    ) -> Result<bool, TufRepoContentsError> {
+        // TODO-correctness: We should return false regardless of `zone_kind` if
+        // there are still pending updates for components earlier in the update
+        // ordering than zones: RoT bootloader / RoT / SP / Host OS.
+
+        match zone_kind {
+            ZoneKind::Nexus => {
+                // Nexus can only be updated if all non-Nexus zones have been
+                // updated, i.e., their image source is an artifact from the new
+                // repo.
+                let new_repo = self.input.tuf_repo().description();
+
+                // If we don't actually have a TUF repo here, we can't do
+                // updates anyway; any return value is fine.
+                if new_repo.tuf_repo().is_none() {
+                    return Ok(false);
+                }
+
+                // Check that all in-service zones (other than Nexus) on all
+                // sleds have an image source consistent with `new_repo`.
+                for sled_id in self.blueprint.sled_ids_with_zones() {
+                    for z in self.blueprint.current_sled_zones(
+                        sled_id,
+                        BlueprintZoneDisposition::is_in_service,
+                    ) {
+                        let kind = z.zone_type.kind();
+                        if kind != ZoneKind::Nexus
+                            && z.image_source
+                                != new_repo.zone_image_source(kind)?
+                        {
+                            return Ok(false);
+                        }
+                    }
+                }
+
+                Ok(true)
+            }
+            _ => Ok(true), // other zone kinds have no special dependencies
+        }
+    }
+
+    /// Return `true` iff we believe a zone can safely be shut down; e.g., any
+    /// data it's responsible for is sufficiently persisted or replicated.
+    ///
+    /// "shut down" includes both "discretionary expunge" (e.g., if we're
+    /// dealing with a zone that is updated via expunge -> replace) or "shut
+    /// down and restart" (e.g., if we're upgrading a zone in place).
+    ///
+    /// This function is not (and cannot!) be called in the "expunge a zone
+    /// because the underlying disk / sled has been expunged" case. In this
+    /// case, we have no choice but to reconcile with the fact that the zone is
+    /// now gone.
+    fn can_zone_be_shut_down_safely(&self, zone_kind: ZoneKind) -> bool {
```

Comment:
Should we pass the whole `zone` object here? I know we don't need it yet, but it seems possible we'd want to actually answer "can this zone UUID terminate" rather than "can an arbitrary zone of this type" terminate.

(Maybe we don't care, I just think it's worth identifying, we're talking about "ANY zone of a particular kind", not necessarily a specific zone - even though, in the calling context, we **are** asking about a single specific zone)

Reply:
Good question. We _have_ the whole zone object, so it seems harmless to pass the whole thing in? Then if we ever need anything about the specific zone, we have it hand. :+1: